/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

option java_package = "org.apache.tajo.ipc";
option java_outer_classname = "ClientProtos";
option java_generic_services = true;
option java_generate_equals_and_hash = true;

import "tajo_protos.proto";
import "TajoIdProtos.proto";
import "CatalogProtos.proto";
import "PrimitiveProtos.proto";

enum ResultCode {
  OK = 0;
  ERROR = 1;
}

message CreateSessionRequest {
  required string username = 1;
  optional string baseDatabaseName = 2;
}

message CreateSessionResponse {
  enum ResultState {
      SUCCESS = 0;
      FAILED = 1;
  }
  required ResultState state = 1;
  optional SessionIdProto sessionId = 2;
  optional string message = 3;
}

message UpdateSessionVariableRequest {
  optional SessionIdProto sessionId = 1;
  optional KeyValueSetProto setVariables = 2;
  repeated string unsetVariables = 3;
}

message SessionedStringProto {
  optional SessionIdProto sessionId = 1;
  required string value = 2;
}

message ExplainQueryResponse {
  required ResultCode resultCode = 1;
  optional string explain = 2;
  optional string errorMessage = 3;
}

message QueryRequest {
  optional SessionIdProto sessionId = 1;
  required string query = 2;
  repeated KeyValueProto setVariables = 3;
}

message UpdateQueryResponse {
  required ResultCode resultCode = 1;
  optional string errorMessage = 2;
}

message GetQueryResultRequest {
  optional SessionIdProto sessionId = 1;
  required QueryIdProto queryId = 2;
}

message GetQueryResultResponse {
  optional TableDescProto tableDesc = 1;
  optional string errorMessage = 2;
  required string tajoUserName = 3;
}

message KillQueryRequest {
  optional SessionIdProto sessionId = 1;
  required QueryIdProto queryId = 2;
}

message GetQueryListRequest {
  optional SessionIdProto sessionId = 1;
}

message BriefQueryInfo {
  required QueryIdProto queryId = 1;
  required QueryState state = 2;
  required int64 startTime = 3;
  required int64 finishTime = 4;
  required string query = 5;
  required string queryMasterHost = 6;
  required int32 queryMasterPort = 7;
  required float progress = 8;
}

message GetQueryListResponse {
  repeated BriefQueryInfo queryList = 1;
}

message GetQueryStatusRequest {
  optional SessionIdProto sessionId = 1;
  required QueryIdProto queryId = 2;
}

message SubmitQueryResponse {
  required ResultCode resultCode = 1;
  required QueryIdProto queryId = 2;
  required string userName = 3;
  optional bool isForwarded = 4 [default = false];

  optional string queryMasterHost = 5;
  optional int32 queryMasterPort = 6;

  message SerializedResultSet {
    optional SchemaProto schema = 1;
    optional int32 bytesNum = 2;
    repeated bytes serializedTuples = 3;
  }

  optional SerializedResultSet resultSet = 7;
  optional TableDescProto tableDesc = 8;
  optional int32 maxRowNum = 9;

  optional string errorMessage = 10;
}

message GetQueryStatusResponse {
  required ResultCode resultCode = 1;
  required QueryIdProto queryId = 2;
  optional QueryState state = 3;
  optional float progress = 4;
  optional int64 submitTime = 5;
  optional int64 finishTime = 7;
  optional bool hasResult = 8;
  optional string errorMessage = 9;
  optional string queryMasterHost = 10;
  optional int32 queryMasterPort = 11;
}

message GetClusterInfoRequest {
  optional SessionIdProto sessionId = 1;
}

message WorkerResourceInfo {
  required string allocatedHost = 1;
  required int32 peerRpcPort = 2;
  required int32 queryMasterPort = 3;
  required int32 clientPort = 4;
  required int32 pullServerPort = 5;
  required int32 httpPort = 6;
  required float diskSlots = 7;
  required int32 cpuCoreSlots = 8;
  required int32 memoryMB = 9;
  required float usedDiskSlots = 10;
  required int32 usedMemoryMB = 11;
  required int32 usedCpuCoreSlots = 12;
  required int64 maxHeap = 13;
  required int64 freeHeap = 14;
  required int64 totalHeap = 15;
  required int32 numRunningTasks = 16;
  required string workerStatus = 17;
  required int64 lastHeartbeat = 18;
  required bool queryMasterMode = 19;
  required bool taskRunnerMode = 20;
  required int32 numQueryMasterTasks = 21;
}

message GetClusterInfoResponse {
  repeated WorkerResourceInfo workerList = 1;
}

message GetTableListRequest {
  optional SessionIdProto sessionId = 1;
  optional string databaseName = 2;
}

message GetTableListResponse {
  repeated string tables = 1;
}

message GetTableDescRequest {
  optional SessionIdProto sessionId = 1;
  required string tableName = 2;
}

message CreateTableRequest {
  optional SessionIdProto sessionId = 1;
  required string name = 2;
  required SchemaProto schema = 3;
  required TableProto meta = 4;
  required string path = 5;
  optional PartitionMethodProto partition = 6;
}

message DropTableRequest {
  optional SessionIdProto sessionId = 1;
  required string name = 2;
  optional bool purge = 3 [default = false];
}

message TableResponse {
  required ResultCode resultCode = 1;
  optional TableDescProto tableDesc = 2;
  optional string errorMessage = 3;
}

message FunctionResponse {
  required ResultCode resultCode = 1;
  repeated FunctionDescProto functions = 2;
  optional string errorMessage = 3;
}
